crates/bootstrap_srv/src/server.rs

```diff
@@ -0,0 +1,333 @@
+
+        // validate agent matches url path
+        if *agent != *info.agent.as_bytes() {
+            return Err(std::io::Error::other("InvalidAgent"));
+        }
+
+        // validate space matches url path
+        if space != info.space {
+            return Err(std::io::Error::other("InvalidSpace"));
+        }
+
+        // validate created at is not older than 3 min ago
+        if info.created_at
+            < now - (std::time::Duration::from_secs(60 * 3).as_micros() as i64)
+        {
+            return Err(std::io::Error::other("InvalidCreatedAt"));
+        }
+
+        // validate created at is less than 3 min in the future
+        if info.created_at
+            > now + (std::time::Duration::from_secs(60 * 3).as_micros() as i64)
+        {
+            return Err(std::io::Error::other("InvalidCreatedAt"));
+        }
+
+        // validate not expired
+        if info.expires_at < now {
+            return Err(std::io::Error::other("InvalidExpiresAt"));
+        }
+
+        // validate expires_at is not before (or equal to) created_at
+        if info.expires_at <= info.created_at {
+            return Err(std::io::Error::other("InvalidExpiresAt"));
+        }
+
+        // validate expires_at is not more than 30 min after created_at
+        if info.expires_at - info.created_at
+            > (std::time::Duration::from_secs(60 * 30).as_micros() as i64)
+        {
+            return Err(std::io::Error::other("InvalidExpiresAt"));
+        }
+
+        // validate signature (do this at the end because it's more expensive
+        info.agent
+            .verify(info.encoded.as_bytes(), &info.signature)
+            .map_err(|err| {
+                std::io::Error::other(format!("InvalidSignature: {err:?}"))
+            })?;
+
+        let r = if info.is_tombstone {
+            None
+        } else {
+            Some(self.store.write(&info_raw)?)
+        };
+
+        self.space_map.update(
+            self.config.max_entries_per_space,
+            space,
+            Some((info, r)),
+        );
+
+        Ok((200, b"{}".to_vec()))
+    }
+
+    /// Helper to get the next path segment as Bytes.
+    fn path_to_bytes(&mut self) -> std::io::Result<bytes::Bytes> {
+        use base64::prelude::*;
+
+        let p = match self.path.pop() {
+            Some(p) => p,
+            None => return Err(std::io::Error::other("InvalidPathSegment")),
+        };
+
+        Ok(bytes::Bytes::copy_from_slice(
+            &BASE64_URL_SAFE_NO_PAD
+                .decode(p)
+                .map_err(std::io::Error::other)?,
+        ))
+    }
+
+    /// Read the body while respecting our max message size.
+    fn read_body(&mut self) -> std::io::Result<Vec<u8>> {
+        // these are the same right now, but *could* be different
+        const MAX_INFO_SIZE: usize = 1024;
+        const READ_BUF_SIZE: usize = 1024;
+
+        let mut buf = [0; READ_BUF_SIZE];
+        let mut out = Vec::new();
+        loop {
```

Comment:
So the idea is that if reading gets interrupted for some reason, it retries? What conditions would cause it to get interrupted? Should we limit the number of retries to avoid something like a disk failure causing infinite looping?

Reply:
This is the fundamental behavior of the core libc read function, carried forward from back in the days when interrupt requests were how processes were parallelized. When you call `write_all`, under the hood it is doing this exact loop check. There is no need to diverge from cannon here.